File: Java/PalindromeLinkedList.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class PalindromeLinkedList {

    public boolean isPalindrome(ListNode head){
        if (head == null || head.next == null) return true;

        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode prev = null, curr = slow;
        while (curr != null) {
            ListNode nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }

        // Compare both halves
        ListNode first = head, second = prev;
        while (second != null) {
            if (first.val != second.val) return false;
            first = first.next;
            second = second.next;
        }

        return true;
    }

    public static void main(String[] args){
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(2, new ListNode(1))));

        PalindromeLinkedList obj = new PalindromeLinkedList();
        boolean result = obj.isPalindrome(head);

        System.out.println("Is Palindrome? " + result);
    }
}

File: LeetCode/Python/PalindromeLinkedList.py
from typing import Optional
from collections import deque
class ListNode:
    def __init__(self,val):
        self.val=val
        self.next=None
        
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        stack=deque()
        slow,fast=head,head
        while fast and fast.next:
            stack.append(slow.val)
            slow=slow.next
            fast=fast.next.next
        if fast:
            slow=slow.next
        while slow:
            top=stack.pop()
            if top != slow.val:
                return False
            slow=slow.next
        return True



def print_ll(head):
    temp=head
    while temp:
        print(temp.val,end=' -> ')
        temp=temp.next
    print(None)
def listToLL(ls):
    head=ListNode(ls[0])
    curr=head
    for i in (ls[1:]):
        curr.next=ListNode(i)
        curr=curr.next
    return head

s=Solution()
head=listToLL([1,2,2,1])
print_ll(head)
print(f'Is palindrome: {s.isPalindrome(head)}')


# Input: head = [1,2,2,1]
# Output: true

# Input: head = [1,2]
# Output: false